tokio/src/process/unix/pidfd_reaper.rs

```diff
@@ -0,0 +1,265 @@
+    ) -> io::Result<()> {
+        SourceFd(&self.as_raw_fd()).register(registry, token, interest)
+    }
+
+    fn reregister(
+        &mut self,
+        registry: &mio::Registry,
+        token: mio::Token,
+        interest: mio::Interest,
+    ) -> io::Result<()> {
+        SourceFd(&self.as_raw_fd()).reregister(registry, token, interest)
+    }
+
+    fn deregister(&mut self, registry: &mio::Registry) -> io::Result<()> {
+        SourceFd(&self.as_raw_fd()).deregister(registry)
+    }
+}
+
+#[derive(Debug)]
+struct PidfdReaperInner<W>
+where
+    W: Unpin,
+{
+    inner: W,
+    pidfd: PollEvented<Pidfd>,
+}
+
+impl<W> Future for PidfdReaperInner<W>
+where
+    W: Wait + Unpin,
+{
+    type Output = io::Result<ExitStatus>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = Pin::into_inner(self);
+
+        ready!(this.pidfd.poll_read_ready(cx))?;
+        Poll::Ready(Ok(this
+            .inner
+            .try_wait()?
+            .expect("pidfd is ready to read, the process should have exited")))
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct PidfdReaper<W: Wait + Send + Sync + Unpin + 'static>(Option<PidfdReaperInner<W>>);
+
+impl<W> Deref for PidfdReaper<W>
+where
+    W: Wait + Send + Sync + Unpin + 'static,
+{
+    type Target = W;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0.as_ref().expect("inner has gone away").inner
+    }
+}
+
+impl<W> PidfdReaper<W>
+where
+    W: Wait + Send + Sync + Unpin + 'static,
+{
+    pub(crate) fn new(inner: W) -> Result<Self, (Option<io::Error>, W)> {
+        if let Some(pidfd) = Pidfd::open(inner.id()) {
+            match PollEvented::new_with_interest(pidfd, Interest::READABLE) {
+                Ok(pidfd) => Ok(Self(Some(PidfdReaperInner { pidfd, inner }))),
+                Err(io_error) => Err((Some(io_error), inner)),
+            }
+        } else {
+            Err((None, inner))
+        }
+    }
+
+    pub(crate) fn inner_mut(&mut self) -> &mut W {
+        &mut self.0.as_mut().expect("inner has gone away").inner
+    }
+}
+
+impl<W> Future for PidfdReaper<W>
+where
+    W: Wait + Send + Sync + Unpin + 'static,
+{
+    type Output = io::Result<ExitStatus>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        Pin::new(
+            Pin::into_inner(self)
+                .0
+                .as_mut()
+                .expect("inner has gone away"),
+        )
+        .poll(cx)
+    }
+}
+
+impl<W> Kill for PidfdReaper<W>
+where
+    W: Wait + Send + Sync + Unpin + Kill + 'static,
+{
+    fn kill(&mut self) -> io::Result<()> {
+        self.inner_mut().kill()
+    }
+}
+
+impl<W> Drop for PidfdReaper<W>
+where
+    W: Wait + Send + Sync + Unpin + 'static,
+{
+    fn drop(&mut self) {
+        let mut reaper_inner = self.0.take().expect("inner has gone away");
+        if let Ok(Some(_)) = reaper_inner.inner.try_wait() {
+            return;
+        }
+
+        Handle {
```

Comment:
Thanks I think adding a new runtime driver is the way to go.

However I can't just store the `PidfdReaper` into a global queue like `GlobalOrphanQueue` does since the `PollEvented` is linked to a runtime, so I will insert the `PidfdReaper` directly into the linked `runtime::process::Driver`.

If the `Driver` is already dropped, then it will fallback to `GlobalOrphanQueue`.

Reply:
Perhaps I should just put the `StdChild` into the `GlobalOrphanQueue` in the `drop` implementation, since this is the first impl/use of pidfd in tokio, I should keep it simple and easy to review and improve on the `GlobalOrphanQueue::reap` efficiency later.